File: libs/java/jmstest/src/test/java/com/broker/itest/jms/queue/QueueBrowserTest.java
/*
 * Copyright 2014-present IVK JSC. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
 
package com.broker.itest.jms.queue;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

import java.util.Enumeration;

import javax.jms.JMSException;
import javax.jms.Message;
import javax.jms.QueueBrowser;
import javax.jms.TextMessage;

import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TestName;

import com.broker.itest.provider.ProviderLoader;
import com.broker.itest.testcase.QueueConnectionTestCase;

/**
 * Test the <code>javax.jms.QueueBrowser</code> features.
 */
public class QueueBrowserTest extends QueueConnectionTestCase {

	@Rule
	public TestName name = new TestName();

	@Override
	public String getTestName() {
		return name.getMethodName();
	}

	/**
	 * The <code>QueueBrowser</code> of the receiver's session
	 */
	protected QueueBrowser receiverBrowser;

	/**
	 * The <code>QueueBrowser</code> of the sender's session
	 */
	protected QueueBrowser senderBrowser;

	/**
	 * Test the <code>QueueBrowser</code> of the sender.
	 */

	@Test
	public void testSenderBrowser() {
		try {
			TextMessage message_1 = senderSession.createTextMessage();
			message_1.setText("testBrowser:message_1");
			TextMessage message_2 = senderSession.createTextMessage();
			message_2.setText("testBrowser:message_2");

			// send two messages...
			sender.send(message_1);
			sender.send(message_2);

			// ask the browser to browse the sender's session
			Enumeration<?> enumeration = senderBrowser.getEnumeration();
			int count = 0;
			while (enumeration.hasMoreElements()) {
				// one more message in the queue
				count++;
				// check that the message in the queue is one of the two which where
				// sent
				Object obj = enumeration.nextElement();
				assertTrue("No message was received", obj != null);
				if (obj == null)
					return;
				assertTrue(obj instanceof TextMessage);
				TextMessage msg = (TextMessage) obj;
				assertTrue(msg.getText().startsWith("testBrowser:message_"));
			}
			// check that there is effectively 2 messages in the queue
			assertEquals(2, count);

			// receive the first message...
			Message m = receiver.receive(ProviderLoader.TIMEOUT);
			// ... and check it is the first which was sent.
			assertTrue(m instanceof TextMessage);
			TextMessage msg = (TextMessage) m;
			assertEquals("testBrowser:message_1", msg.getText());

			// receive the second message...
			m = receiver.receive(ProviderLoader.TIMEOUT);
			// ... and check it is the second which was sent.
			assertTrue(m instanceof TextMessage);
			msg = (TextMessage) m;
			assertEquals("testBrowser:message_2", msg.getText());

			// ask the browser to browse the sender's session
			enumeration = receiverBrowser.getEnumeration();
			// check that there is no messages in the queue
			// (the two messages have been acknowledged and so removed
			// from the queue)
			assertTrue(!enumeration.hasMoreElements());
		} catch (JMSException e) {
			fail(e.getMessage());
		}
	}

	/**
	 * Test the <code>QueueBrowser</code> after receive.
	 */

	@Test
	public void testSendBrowserReceiveBrowser() {
		try {
			TextMessage message = senderSession.createTextMessage();
			message.setText("testSendBrowserReceiveBrowser");

			int i = 0;
			while (i != 20) {
				message.setIntProperty("testSendBrowserReceiveBrowser", i);
				sender.send(message);
				i++;
			}

			Enumeration<?> enumerationSender = senderBrowser.getEnumeration();
			i = 0;
			while (enumerationSender.hasMoreElements()) {
				Object obj = enumerationSender.nextElement();
				assertTrue("No message was received", obj != null);
				if (obj == null)
					return;
				assertTrue(obj instanceof TextMessage);
				TextMessage msg = (TextMessage) obj;
				assertTrue(msg.getText().equals("testSendBrowserReceiveBrowser"));
				assertTrue(msg.getIntProperty("testSendBrowserReceiveBrowser") == i);
				i++;
			}
			assertEquals(i, 20);

			i = 0;
			while (i != 10) {
				Message m = receiver.receive(ProviderLoader.TIMEOUT);
				assertTrue(m instanceof TextMessage);
				TextMessage msg = (TextMessage) m;
				assertTrue(msg.getText().equals("testSendBrowserReceiveBrowser"));
				assertTrue(msg.getIntProperty("testSendBrowserReceiveBrowser") == i);
				i++;
			}

			Enumeration<?> enumerationReceiver = receiverBrowser.getEnumeration();
			i = 0;
			while (enumerationReceiver.hasMoreElements()) {
				Object obj = enumerationReceiver.nextElement();
				assertTrue("No message was received", obj != null);
				if (obj == null)
					return;
				assertTrue(obj instanceof TextMessage);
				TextMessage msg = (TextMessage) obj;
				assertTrue(msg.getText().equals("testSendBrowserReceiveBrowser"));
				assertTrue((msg.getIntProperty("testSendBrowserReceiveBrowser") - 10) == i);
				i++;
			}
			assertEquals(i, 10);

		} catch (JMSException e) {
			fail(e.getMessage());
		}
	}

	/**
	 * Test that a <code>QueueBrowser</cdeo> created with a message selector browses only the messages matching this
	 * selector.
	 */
	@Test
	public void testBrowserWithMessageSelector() {
		try {
			senderBrowser = senderSession.createBrowser(senderQueue, "pi = 3.14159");

			TextMessage message_1 = senderSession.createTextMessage();
			message_1.setText("testBrowserWithMessageSelector:message_1");
			TextMessage message_2 = senderSession.createTextMessage();
			message_2.setDoubleProperty("pi", 3.14159);
			message_2.setText("testBrowserWithMessageSelector:message_2");

			sender.send(message_1);
			sender.send(message_2);

			Enumeration<?> enumeration = senderBrowser.getEnumeration();
			int count = 0;
			while (enumeration.hasMoreElements()) {
				count++;
				Object obj = enumeration.nextElement();
				assertTrue("No message was received", obj != null);
				if (obj == null)
					return;
				assertTrue(obj instanceof TextMessage);
				TextMessage msg = (TextMessage) obj;
				assertEquals("testBrowserWithMessageSelector:message_2", msg.getText());
			}
			assertEquals(1, count);
		} catch (JMSException e) {
			fail(e.getMessage());
		}
	}

	@Override
	public void setUp() {
		try {
			super.setUp();
			receiverBrowser = receiverSession.createBrowser(receiverQueue);
			senderBrowser = senderSession.createBrowser(senderQueue);
		} catch (JMSException e) {
			e.printStackTrace();
		}
	}

	@Override
	public void tearDown() {
		try {
			if (receiverBrowser != null)
				receiverBrowser.close();
			if (senderBrowser != null)
				senderBrowser.close();
			super.tearDown();
		} catch (JMSException e) {
			e.printStackTrace();
		} finally {
			receiverBrowser = null;
			senderBrowser = null;
		}
	}
}

File: tests/brokertest/QueueBrowserTest.cpp
/*
 * Copyright 2014-present IVK JSC. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "QueueBrowserTest.h"
#include <sstream>
#include <fake_cpp14.h>

////////////////////////////////////////////////////////////////////////////////
void QueueBrowserTest::SetUp() {
  cmsProvider = std::make_unique<CMSProvider>(getBrokerURL());
  cmsProvider->cleanUpDestination();
}
////////////////////////////////////////////////////////////////////////////////
TEST_F(QueueBrowserTest, testReceiveBrowseReceive) {
  cms::Session *session(cmsProvider->getSession());

  std::unique_ptr<cms::Queue> queue(session->createQueue(::testing::UnitTest::GetInstance()->current_test_info()->name()));
  std::unique_ptr<cms::MessageConsumer> consumer(session->createConsumer(queue.get()));
  std::unique_ptr<cms::MessageProducer> producer(session->createProducer(queue.get()));

  std::unique_ptr<cms::TextMessage> message1(session->createTextMessage("First Message"));
  std::unique_ptr<cms::TextMessage> message2(session->createTextMessage("Second Message"));
  std::unique_ptr<cms::TextMessage> message3(session->createTextMessage("Third Message"));

  EXPECT_NO_THROW(producer->send(message1.get()));
  EXPECT_NO_THROW(producer->send(message2.get()));
  EXPECT_NO_THROW(producer->send(message3.get()));
  EXPECT_NO_THROW(producer->close());

  // Get the first.
  std::unique_ptr<cms::TextMessage> inbound;
  EXPECT_NO_THROW(inbound.reset(dynamic_cast<cms::TextMessage *>(consumer->receive(3000))));
  EXPECT_TRUE(inbound != nullptr);
  message1->setReadable();
  EXPECT_EQ(message1->getText(), inbound->getText());

  EXPECT_NO_THROW(consumer->close());

  std::unique_ptr<cms::QueueBrowser> browser;
  EXPECT_NO_THROW(browser.reset(session->createBrowser(queue.get())));
  cms::MessageEnumeration *enumeration = nullptr;
  EXPECT_NO_THROW(enumeration = browser->getEnumeration());
  EXPECT_TRUE(enumeration != nullptr);
  // browse the second
  EXPECT_TRUE(enumeration->hasMoreMessages()) << "should have received the second message";
  EXPECT_NO_THROW(inbound.reset(dynamic_cast<cms::TextMessage *>(enumeration->nextMessage())));
  EXPECT_TRUE(inbound != nullptr);
  message2->setReadable();
  EXPECT_EQ(message2->getText(), inbound->getText());

  // browse the third.
  EXPECT_TRUE(enumeration->hasMoreMessages()) << "should have received the third message";
  EXPECT_NO_THROW(inbound.reset(dynamic_cast<cms::TextMessage *>(enumeration->nextMessage())));
  EXPECT_TRUE(inbound != nullptr);
  message3->setReadable();
  EXPECT_EQ(message3->getText(), inbound->getText());

  // There should be no more.
  bool tooMany = false;
  while (enumeration->hasMoreMessages()) {
    tooMany = true;
  }
  EXPECT_TRUE(!tooMany) << "should not have browsed any more messages";
  EXPECT_NO_THROW(browser->close());

  EXPECT_NO_THROW(browser.reset(session->createBrowser(queue.get())));
  EXPECT_NO_THROW(enumeration = browser->getEnumeration());
  // browse the second
  EXPECT_TRUE(enumeration->hasMoreMessages()) << "should have received the second message";
  EXPECT_NO_THROW(inbound.reset(dynamic_cast<cms::TextMessage *>(enumeration->nextMessage())));
  EXPECT_TRUE(inbound != nullptr);
  EXPECT_EQ(message2->getText(), inbound->getText());
  // browse the third.
  EXPECT_TRUE(enumeration->hasMoreMessages()) << "should have received the third message";
  EXPECT_NO_THROW(inbound.reset(dynamic_cast<cms::TextMessage *>(enumeration->nextMessage())));
  EXPECT_TRUE(inbound != nullptr);
  EXPECT_EQ(message3->getText(), inbound->getText());

  // Re-open the consumer
  EXPECT_NO_THROW(consumer.reset(session->createConsumer(queue.get())));
  // Receive the second.
  EXPECT_NO_THROW(inbound.reset(dynamic_cast<cms::TextMessage *>(consumer->receive(3000))));
  EXPECT_TRUE(inbound != nullptr);
  EXPECT_EQ(message2->getText(), inbound->getText());
  // Receive the third.
  EXPECT_NO_THROW(inbound.reset(dynamic_cast<cms::TextMessage *>(consumer->receive(3000))));
  EXPECT_TRUE(inbound != nullptr);
  EXPECT_EQ(message3->getText(), inbound->getText());

  EXPECT_NO_THROW(consumer->close());
  EXPECT_NO_THROW(browser->close());
}

////////////////////////////////////////////////////////////////////////////////
TEST_F(QueueBrowserTest, testBrowseReceive) {
  std::unique_ptr<cms::TextMessage> inbound;

  cms::Session *session(cmsProvider->getSession());

  std::unique_ptr<cms::Queue> queue(session->createQueue(::testing::UnitTest::GetInstance()->current_test_info()->name()));
  std::unique_ptr<cms::MessageProducer> producer(session->createProducer(queue.get()));
  std::unique_ptr<cms::TextMessage> message1(session->createTextMessage("First Message"));

  producer->send(message1.get());

  // create browser first
  std::unique_ptr<cms::QueueBrowser> browser(session->createBrowser(queue.get()));
  cms::MessageEnumeration *enumeration = browser->getEnumeration();

  // create consumer
  std::unique_ptr<cms::MessageConsumer> consumer(session->createConsumer(queue.get()));

  // browse the first message
  EXPECT_TRUE(enumeration->hasMoreMessages()) << "should have received the first message";
  inbound.reset(dynamic_cast<cms::TextMessage *>(enumeration->nextMessage()));
  EXPECT_TRUE(inbound.get() != nullptr);
  message1->setReadable();
  EXPECT_EQ(message1->getText(), inbound->getText());

  // Receive the first message.
  inbound.reset(dynamic_cast<cms::TextMessage *>(consumer->receive(3000)));
  EXPECT_TRUE(inbound.get() != nullptr);
  EXPECT_EQ(message1->getText(), inbound->getText());

  consumer->close();
  browser->close();
  producer->close();
}

////////////////////////////////////////////////////////////////////////////////
TEST_F(QueueBrowserTest, testQueueBrowserWith2Consumers) {
  std::stringstream sstr;

  static const int numMessages = 5;

  Connection *connection = cmsProvider->getConnection();

  EXPECT_TRUE(connection != nullptr);

  std::unique_ptr<cms::Session> session(connection->createSession(cms::Session::CLIENT_ACKNOWLEDGE));

  std::unique_ptr<cms::Queue> queue(
      session->createQueue(::testing::UnitTest::GetInstance()->current_test_info()->name()));  // testQueueBrowserWith2Consumers
  std::unique_ptr<cms::Queue> queuePrefetch10(session->createQueue(
      ::testing::UnitTest::GetInstance()->current_test_info()->name()));  // testQueueBrowserWith2Consumers?consumer.prefetchSize=10
  std::unique_ptr<cms::Queue> queuePrefetch1(session->createQueue(
      ::testing::UnitTest::GetInstance()->current_test_info()->name()));  // testQueueBrowserWith2Consumers?consumer.prefetchSize=1

  std::unique_ptr<ConnectionFactory> factory(cms::ConnectionFactory::createCMSConnectionFactory(cmsProvider->getBrokerURL()));
  std::unique_ptr<Connection> connection2(factory->createConnection());

  connection2->start();

  std::unique_ptr<cms::Session> session2(connection2->createSession(cms::Session::AUTO_ACKNOWLEDGE));

  std::unique_ptr<cms::MessageProducer> producer(session->createProducer(queue.get()));
  std::unique_ptr<cms::MessageConsumer> consumer(session->createConsumer(queuePrefetch10.get()));

  producer->setDeliveryMode(cms::DeliveryMode::NON_PERSISTENT);

  for (int i = 0; i < numMessages; i++) {
    sstr.str("");
    sstr << "Message: " << i;
    std::unique_ptr<cms::TextMessage> message(session->createTextMessage(sstr.str()));
    producer->send(message.get());
  }

  std::unique_ptr<cms::QueueBrowser> browser(session2->createBrowser(queuePrefetch1.get()));
  cms::MessageEnumeration *browserView = browser->getEnumeration();

  std::vector<std::unique_ptr<cms::Message>> messages;
  for (int i = 0; i < numMessages; i++) {
    std::unique_ptr<cms::Message> m1(consumer->receive(3000));
    sstr.str("");
    sstr << "m1 is null for index: " << i;
    EXPECT_TRUE(m1 != nullptr) << sstr.str();
    messages.emplace_back(std::move(m1));
  }

  for (int i = 0; i < numMessages && browserView->hasMoreMessages(); i++) {
    cms::Message *m1 = messages[i].get();
    std::unique_ptr<cms::Message> m2(browserView->nextMessage());
    sstr.str("");
    sstr << "m2 is null for index: " << i;
    EXPECT_TRUE(m2 != nullptr) << sstr.str();
    EXPECT_EQ(m1->getCMSMessageID(), m2->getCMSMessageID());
  }

  EXPECT_FALSE(browserView->hasMoreMessages()) << "nothing left in the browser";
  std::unique_ptr<cms::TextMessage> messageEmpty;
  messageEmpty.reset(dynamic_cast<TextMessage *>(consumer->receiveNoWait()));
  EXPECT_TRUE(messageEmpty == nullptr) << "expect empty message, but got " << messageEmpty->getText() << " [" << messageEmpty->getCMSMessageID()
                                       << "]";

  for (auto &msg : messages) {
    EXPECT_NO_THROW(msg->acknowledge());
  }
  messages.clear();

  browser->close();
  producer->close();
  consumer->close();
  session->close();
  session2->close();
}

void QueueBrowserTest::TearDown() {}
